bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/CaddxProperty.java

```diff
@@ -0,0 +1,128 @@
+ */
+package org.openhab.binding.caddx.internal;
+
+import java.util.Arrays;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * Panel message property class
+ *
+ * @author Georgios Moutsos - Initial contribution
+ */
+@NonNullByDefault
+public class CaddxProperty {
+    // private
+    private final String name;
+    private final CaddxPropertyType type; // 'Int', 'String', 'Bit'
+    private final int byteFrom;
+    private final int byteLength;
+    private final int bitFrom;
+    private final int bitLength;
+    private final boolean external;
+    private final String id;
+
+    // Constructor
+    public CaddxProperty(String id, int byteFrom, int byteLength, int bitFrom, int bitLength, CaddxPropertyType type,
+            String name, boolean external) {
+        this.id = id;
+        this.name = name;
+        this.type = type;
+        this.byteFrom = byteFrom;
+        this.byteLength = byteLength;
+        this.bitFrom = bitFrom;
+        this.bitLength = bitLength;
+        this.external = external;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public CaddxPropertyType getType() {
+        return type;
+    }
+
+    public boolean getExternal() {
+        return external;
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    public String getValue(byte[] message) {
+        int mask;
+        int val;
+
+        switch (type) {
+            case INT:
+                if (bitFrom == 0 && bitLength == 0) {
+                    mask = 255;
+                    val = message[byteFrom - 1] & mask;
+                } else {
+                    mask = ((1 << ((bitLength - bitFrom))) - 1) << bitFrom;
+                    val = (message[byteFrom - 1] & mask) >> bitFrom;
+                }
+
+                return Integer.toString(val);
+            case STRING:
+                byte[] str = Arrays.copyOfRange(message, byteFrom - 1, byteFrom + byteLength);
+                return new String(str);
+            case BIT:
+                return (((message[byteFrom - 1] & (1 << bitFrom)) > 0) ? "true" : "false");
+            default:
+                throw new IllegalArgumentException("type is unknown.");
+        }
+    }
+
+    public String toString(byte[] message) {
+        int mask;
+        int val;
+
+        switch (type) {
+            case INT:
+                if (bitFrom == 0 && bitLength == 0) {
+                    mask = 255;
+                    val = message[byteFrom - 1];
+                } else {
+                    mask = ((1 << ((bitLength - bitFrom) + 1)) - 1) << bitFrom;
+                    val = (message[byteFrom - 1] & mask) >> bitFrom;
+                }
+
+                return name + ": " + String.format("%2s", Integer.toHexString(val)) + " - " + Integer.toString(val)
+                        + " - " + ((val >= 32 && val <= 'z') ? ((char) val) : "-");
+            case STRING:
+                StringBuilder sb = new StringBuilder();
```

Comment:
It might be more appropriate to wrap a StringWriter with a PrintWriter and use the PrintWriter instead of the StringBuilder. With the PrintWriter you get `println` and `printf`, both of which you could use here.

Reply:
Done